src/gc/collector.cpp

```diff
@@ -342,26 +416,121 @@ void markPhase() {
 #endif
 }
 
+void finalizationOrderingFirstPass(Box* obj) {
+    std::shared_ptr<FirstPhaseStack> stack = std::make_shared<FirstPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        visitByGCKind(p, visitor);
+    }
+}
+
+void finalizationOrderingSecondPass(Box* obj) {
+    std::shared_ptr<SecondPhaseStack> stack = std::make_shared<SecondPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        GCAllocation* al = GCAllocation::fromUserData(p);
+        assert(orderingState(al) != FinalizationState::UNREACHABLE);
+        visitByGCKind(p, visitor);
+    }
+}
+
+// Implementation of PyPy's finalization ordering algorithm:
+// http://pypy.readthedocs.org/en/latest/discussion/finalizer-order.html
+void finalizationOrderingPhase() {
+    // TODO: Replace with iterator or something that takes a lambda for memory efficiency?.
+    std::vector<Box*> objects_with_finalizers;
+    global_heap.getOrderedFinalizers(objects_with_finalizers);
+
+    std::vector<Box*> finalizer_marked;
+
+    for (Box* obj : objects_with_finalizers) {
+        GCAllocation* al = GCAllocation::fromUserData(obj);
+
+        // We are only interested in object with finalizers that need to be
+        // garbage-collected.
+        if (orderingState(al) == FinalizationState::UNREACHABLE) {
+            // Unordered finalizers don't block the objects they reference from being finalized.
+            assert(hasOrderedFinalizer(obj));
+            finalizer_marked.push_back(obj);
+            finalizationOrderingFirstPass(obj);
+            finalizationOrderingSecondPass(obj);
+        }
+    }
+
+    for (Box* marked : finalizer_marked) {
+        GCAllocation* al = GCAllocation::fromUserData(marked);
+
+        FinalizationState state = orderingState(al);
+        assert(state == FinalizationState::REACHABLE_FROM_FINALIZER || state == FinalizationState::ALIVE);
+
+        if (state == FinalizationState::REACHABLE_FROM_FINALIZER) {
+            assert(roots.count(marked) == 0);
+            registerPermanentRoot(marked);
```

Comment:
I wouldn't use `registerPermanentRoot`/`deregisterPermanentRoot` for the objects in the `pending_finalization_list`.  instead just add a pass iterating the list during the markPhase.